### DISCUSSION TITLE
Problem to set instance as Default Search Engine in Firefox

### DISCUSSION THREAD

**Discussion by jacktheripper19** [searxng/searxng#1288] [Q&A] [answered]
Hello,
I deployed a searxng instance as a docker container and it works well. However, I can't set it as a default search engine in firefox: I'm using firefox version 101. I followed the guide by checking the search/address bar but I never got the option to add to search engines. 
I thought that this was an issue from my instance configs but I can though set it as default for chrome-based browsers and even on mobile app of firefox. Then I started thinking that maybe the issue is from the browser itself but I can add other instances though. So I checked the list of public instances and it seems that I have the same issue with some of them.
For example, I can add https://searx.tiekoetter.com/ but not https://searx.loafland.xyz/
I wonder if there is some setting that I missed or if there is a way to check where the issue is coming from.

**Comment by return42** [comment]
FWI I use FFox 100.0.2 and I don't have any issue with [OpenSearch](https://github.com/searxng/searxng/issues/431) / I tested with tiekoetter's instance.

**Reply by jacktheripper19** [reply]
Yes I can do that too. But what about [loaflan](https://searx.loafland.xyz/) ?

**Reply by return42** [reply]
Same; https://searx.loafland.xyz/ --> FFox 100.0.2 and I don't have any issue with [OpenSearch](https://github.com/searxng/searxng/issues/431)

**Comment by jacktheripper19** [comment, marked as answer]
Appearently, on firefox, if there is an searXNG instance that was already added as a search engine you can't add a different instance having the same display name "searXNG". Therefore I solved my problem by deleting the public instance that I added before. Then I was able t add my own instance.

**Reply by dalf** [reply]
Perhaps it can be solve with a pull request:
https://github.com/searxng/searxng/blob/ea0cddba0b9366326eaebc0a3a55502a17e5baf4/searx/settings_defaults.py#L142

Here we change the default to `None`, and then in `settings_set_defaults` we add:
```python
if settings['general']['instance_name'] is None and settings['server']['base_url']:
    settings['general']['instance_name'] = f"SearXNG - { urlparse(settings['server']['base_url']).hostname }"
```

So the two `opensearch*.xml` endpoints are going to reference that:
https://github.com/searxng/searxng/blob/ea0cddba0b9366326eaebc0a3a55502a17e5baf4/searx/templates/simple/opensearch.xml#L3

So multiple instances are going to be seen as different search engines by the browsers.

---

It won't solve the issue for the instances with a value for BASE_URL, so may we can just patch change `<ShortName>{{ instance_name }}</ShortName>` into `<ShortName>{{ instance_name or base_url_hostname }}</ShortName>` (`base_url_hostname` defined using the Flask request).

**Reply by jacktheripper19** [reply]
I agree. This way, one doesn't have to keep only one instance in the list of engines. Although I think it is an issue from firefox's implementation as it should check if the search engine to add already exists in the list by using its url and not the dispay name.

**Reply by GGORG0** [reply]
Had the same issue with LibreWolf (includes https://searx.be as SearXNG by default) here https://github.com/searxng/searxng-docker/issues/112#issuecomment-1374784676

**Comment by rasos** [comment]
When trying to add our instance in Firefox 147.0.3 on Xubuntu  in the URL bar with a right mouse click, I am getting: 

> Firefox could not install the search engine from: https://fairsuch.net/opensearch.xml?method=POST&autocomplete=startpage

 The Firefox has just been cleaned, so there is no other searxng instance. 
 
It works when adding https://fairsuch.net/search?q=%s manually. 
 
opensearch.xml looks okay: 

```
 <?xml version="1.0" encoding="utf-8"?>
<OpenSearchDescription xmlns="http://a9.com/-/spec/opensearch/1.1/" xmlns:moz="http://www.mozilla.org/2006/browser/search/">
  <ShortName>fairsuch</ShortName>
  <LongName>SearXNG metasearch</LongName>
  <Description>SearXNG is a metasearch engine that respects your privacy.</Description>
  <InputEncoding>UTF-8</InputEncoding>
  <Image type="image/png">/static/themes/simple/img/favicon.png</Image>
  <Url rel="results" type="text/html" method="POST" template="https://fairsuch.net/search">
    <Param name="q" value="{searchTerms}" />
  </Url>
    <Url rel="suggestions" type="application/x-suggestions+json" method="POST" template="https://fairsuch.net/autocompleter?q={searchTerms}"/>
  <Url rel="self" type="application/opensearchdescription+xml"  method="POST" template="https://fairsuch.net/opensearch.xml" />
  <Query role="example" searchTerms="SearXNG" />
  <moz:SearchForm>https://fairsuch.net/search</moz:SearchForm>
</OpenSearchDescription>
```
